### PR TITLE
Adding a block height threshold to re-request missing collections 

### PR DESCRIPTION
**Pull request by vishalchangrani**
Fixes: https://github.com/onflow/flow-go/issues/1623

### CODE REVIEW

File: engine/access/ingestion/engine.go [hunk 1]
```diff
@@ -33,12 +33,18 @@ const collectionCatchupDBPollInterval = 10 * time.Millisecond
...
 
+// a threshold of block height beyond which collections should be re-requested (regardless of the number of blocks for which collection are missing)
+// this is to ensure that if a collection is missing for a long time (in terms of block height) it is eventually re-requested
+const missingCollsForHeightThreshold = 100
+
 var defaultCollectionCatchupTimeout = collectionCatchupTimeout
 var defaultCollectionCatchupDBPollInterval = collectionCatchupDBPollInterval
 var defaultFullBlockUpdateInterval = fullBlockUpdateInterval
 var defaultMissingCollsForBlkThreshold = missingCollsForBlkThreshold
+var defaultMissingCollsForHeightThreshold = missingCollsForHeightThreshold
```
Comment (jordanschalm) [inline engine/access/ingestion/engine.go:47]: These seem like they should be config fields on the `Engine` instance rather than global variables

**Review by Kay-Zee — APPROVED** [review]

File: engine/access/ingestion/engine.go (hunk 1 above)
Comment (vishalchangrani) [reply engine/access/ingestion/engine.go:47]: so these private fields were meant to help white box testing where they can be modified as part of the unit test. Yeah, I could add it to the engine too and make them configurable (but may do that as a different issue).

File: engine/access/ingestion/engine.go [hunk 2]
```diff
@@ -33,12 +33,18 @@ const collectionCatchupDBPollInterval = 10 * time.Millisecond
 const fullBlockUpdateInterval = 1 * time.Minute
 
 // a threshold of number of blocks with missing collections beyond which collections should be re-requested
+// this is to prevent spamming the collection nodes with request
 const missingCollsForBlkThreshold = 100
 
+// a threshold of block height beyond which collections should be re-requested (regardless of the number of blocks for which collection are missing)
+// this is to ensure that if a collection is missing for a long time (in terms of block height) it is eventually re-requested
+const missingCollsForHeightThreshold = 100
```
Comment (smnzhu) [inline engine/access/ingestion/engine.go:41]: a better name would be something like `missingCollAgeThreshold`

File: engine/access/ingestion/engine.go [hunk 3]
```diff
@@ -569,8 +575,8 @@ func (e *Engine) updateLastFullBlockReceivedIndex() {
 		}
 	}
 
-	// additionally, if more than threshold blocks have missing collection, re-request those collections
-	if incompleteBlksCnt >= defaultMissingCollsForBlkThreshold {
+	// additionally, if more than threshold blocks have missing collection OR collections are missing since defaultMissingCollsForHeightThreshold, re-request those collections
```
Comment (smnzhu) [inline engine/access/ingestion/engine.go:578]:
```suggestion
	// additionally, if more than threshold blocks have missing collection OR collections are missing for more than defaultMissingCollAgeThreshold, re-request those collections
```

File: engine/access/ingestion/engine_test.go [hunk 4]
```diff
@@ -540,15 +540,46 @@ func (suite *Suite) TestUpdateLastFullBlockReceivedIndex() {
 		// assert that missing collections are requested
 		suite.request.AssertExpectations(suite.T())
 
+		// last full blk index is not advanced
+		suite.blocks.AssertExpectations(suite.T()) // no new call to UpdateLastFullBlockHeight should be made
+	})
+
+	suite.Run("missing collections are requested when count exceeds defaultMissingCollsForHeightThreshold", func() {
```
Comment (smnzhu) [inline engine/access/ingestion/engine_test.go:547]:
```suggestion
	suite.Run("missing collections are requested when age exceeds defaultMissingCollAgeThreshold", func() {
```

File: engine/access/ingestion/engine.go (hunk 2 above)
Comment (vishalchangrani) [reply engine/access/ingestion/engine.go]: changed to `missingCollAgeThreshold`

**Pull request merged**